[navigation-refactor] Use minimal action to navigate inside RHP (#21067)
Expensify/App: contributingGuides/NAVIGATION.md

- To keep the navigation state that was present before changing the layout, we save the state on every change and use it for the `initialState` prop.
Changing the layout means that every component inside `NavigationContainer` is mounted anew.

## Why we need to use minimal action in the `linkTo` function

### The problem 
Let's assume that the user wants to navigate like this:

```
1. Settings_root - navigate > Profile
2. Profile - UP > Settings_root
3. Settings_root - navigate > About 
4. About - browser back button > Settings_root
```

Without minimal action, expected behavior won't be achieved and the final screen will be `Profile`.

Broken behavior is the outcome of two things:
1. Back button in the browser resets the navigation state with the state saved in step two.
   

> 2. `Navigation.navigate` creates action with `getActionFromState` dispatched high in the hierarchy. 

Reviewer: ```suggestion
2. `Navigation.navigate` creates action with `getActionFromState` dispatched at the top level of the navigation hierarchy. 
```

> The reason why `getActionFromState` provided by `react-navigation` is dispatched high in hierarchy is that it doesn't know about current navigation state, only about desired one.

Reviewer: ```suggestion
The reason why `getActionFromState` provided by `react-navigation` is dispatched at the top level of the navigation hierarchy is that it doesn't know about current navigation state, only about desired one.
```

In this example it doesn't know if the `RightModalNavigator` and `Settings` are already mounted.

The action for the first step looks like that: 
```json
{
    "type": "NAVIGATE",
    "payload": {
        "name": "RightModalNavigator",
        "params": {
            "initial": true,
            "screen": "Settings",
            "params": {
                "initial": true,
                "screen": "Profile",
                "path": "/settings/profile"
            }
        }
    }
}
```

> That means, the params for the `RightModalNavigator` and `Settings` (also a navigator) will be filled with the information that we want to navigate to the `Profile` screen.

Reviewer: Could you specify which part contains the information about wanting to navigate to `Profile`? The `path` param?

This information will stay here even if we pop the `Profile` screen and navigate to `About` screen. 

> Later on, when the user presses the browser back button expecting that the `Settings_root` screen will appear, the navigation state will be reset with information about the `Profile` screen in the params and this will be used as a source of truth for the navigation. 

Reviewer: Can you show the code that utilizes this information for more context?

### The solution

If we can create simple action that will only push one screen to the existing navigator, we won't fill any params of the navigators. 

> The `getMinimalAction` compares action generated by the `getActionFromState` with the current navigation state and tries to find a simpler action.

Reviewer: ```suggestion
The `getMinimalAction` compares action generated by the `getActionFromState` with the current navigation state and tries to find the smallest action possible.
```
